libmscore/figuredbass.cpp

```diff
@@ -1770,6 +1773,256 @@ FiguredBass* Score::addFiguredBass()
+            while ( n < 0)
+                  n += 12;
+            return n;
+      };
+      auto get_digit = [](int digit) {
+            // a modifier with no digit always references the 3rd
+            return (digit == FBIDigitNone ) ? 3 : digit;
+      };
+      auto render_harmony = [&] (FiguredBassItem::Modifier prefix, int relpitch, FiguredBassItem::Modifier suffix) {
+            relpitch = get_digit(relpitch);
+            if ( prefix != none && suffix != none) {
+                  qDebug("(1242) measure #%d, cannot render FiguredBassModifier with two modifiers", segment->measure()->no());
+                  return;
+            }
+            FiguredBassItem::Modifier modifier = (prefix == none) ? suffix : prefix;
+            ClefType clef = chord->staff()->clef(segment->tick());
+            
+            // calculate native pitch for line adjusted for transposing instrument.
+            // in doing so we have to SUBTRACT(relpitch-1) because lines are oriented from top to bottom, and
+            // pitches are oriented from bottom to top.  The 1 is because chord harmonies are noted as 1,3,5 (for a normal triad)
+            // rather than (0,2,4).
+            int harmony_line = note->line() - relpitch + 1;
+            int native_pitch = line2pitch(harmony_line, clef, Key::C) + note->ppitch() - note->epitch();
+
+            AccidentalVal acciv = segment->measure()->findAccidental(segment, chord->staff()->idx(), harmony_line);
+            int harmony;
+            switch ( modifier ) {
+                  case FiguredBassItem::Modifier::NONE: {
+                        Note *antecedent = nullptr;
+                        int antecedent_tick = -1; // if we fail to find an antecedence, this will remain -1.
+                        // Here we must apply the accidental seen earlier in the measure if there is one, even if in a different octave.
+                        // find the note with maximum tick, but <= note->tick() whose line % 7 == native_pitch->line%7, take its accidental.
+                        chord->score()->visitNotesInStaff(chord->measure(), [&](Note *n) {
+
+                              if ( n->track() != chord->track())
+                                    return;
+
+                              // we have to skip notes which are not the bottom note of the chord.
+                              // why? because these notes were auto-generated, and their note->line() returns 0 even if not really on line 0.
+                              if ( find_if(n->chord()->notes().begin(),
+                                           n->chord()->notes().end(),
+                                           [n](Note *j) {
+                                                 // if there is a note j whose pitch is < n's pitch, then n is not the bottom note, so return.
+                                                 return j->pitch() < n->pitch();
+                                           }) != n->chord()->notes().end())
```

Comment:
The use of visitNotesInStaff is just temporary.   When I started developing this capability, I wasn’t sure which notes I needed to visit.
In the mean time, the use of visitNotesInStaff is no longer necessary.  I’ve removed the usage from the code in my workarea.
I’ve uploaded a refactored version (see 324d8a34f5c1b8156619bc3207a060492efaa972), as it has the same effect, but you should 
find it much easier to follow the actual code.  Let me know if the new version helps?

https://github.com/jimka2001/MuseScore/commit/324d8a34f5c1b8156619bc3207a060492efaa972 <https://github.com/jimka2001/MuseScore/commit/324d8a34f5c1b8156619bc3207a060492efaa972>


> On 24 May 2015, at 04:53, Maurizio M. Gavioli <notifications@github.com> wrote:
> 
> In libmscore/figuredbass.cpp <https://github.com/musescore/MuseScore/pull/2015#discussion_r30949740>:
> 
> > +                        int antecedent_tick = -1; // if we fail to find an antecedence, this will remain -1.
> > +                        // Here we must apply the accidental seen earlier in the measure if there is one, even if in a different octave.
> > +                        // find the note with maximum tick, but <= note->tick() whose line % 7 == native_pitch->line%7, take its accidental.
> > +                        chord->score()->visitNotesInStaff(chord->measure(), [&](Note *n) {
> > +
> > +                              if ( n->track() != chord->track())
> > +                                    return;
> > +
> > +                              // we have to skip notes which are not the bottom note of the chord.
> > +                              // why? because these notes were auto-generated, and their note->line() returns 0 even if not really on line 0.
> > +                              if ( find_if(n->chord()->notes().begin(),
> > +                                           n->chord()->notes().end(),
> > +                                           [n](Note *j) {
> > +                                                 // if there is a note j whose pitch is < n's pitch, then n is not the bottom note, so return.
> > +                                                 return j->pitch() < n->pitch();
> > +                                           }) != n->chord()->notes().end())
> I have lost you...
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/musescore/MuseScore/pull/2015/files#r30949740>.
>